sat/cli/bootprep/input/base.py

```diff
@@ -38,23 +38,109 @@
 from sat.util import pester_choices
 
 LOGGER = logging.getLogger(__name__)
+inflector = engine()
+
+
+def provides_context(context_var=None):
+    """Get decorator for instance methods which provide Jinja2 context.
+
+    This should be used to decorate instance methods which provide additional
+    context which should be used when rendering Jinja2 templates with the
+    jinja_rendered decorator defined below.
+
+    Note that the context provided by the instance method decorated with this
+    decorator will not be available until that method is called.
+
+    This decorator requires that the instance has the following attribute:
+
+        jinja_context (dict): additional context to use when rendering the
+            result of the method as a Jinja2 template. Defaults to the empty
+            dict if not set.
+    """
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(self, *args, **kwargs):
+            result = func(self, *args, **kwargs)
+            self.jinja_context[context_var or func.__name__] = result
+            return result
+
+        return wrapper
+
+    return decorator
+
+
+def jinja_rendered(func):
```

Comment:
Substituting the `recipe` variable in templates fails with a "variable undefined" error when the `--recipe-version` argument isn't passed, which in my opinion doesn't give much intuition as to what's causing the issue -- when I encountered it, I assumed it was a bug in SAT, not an error in usage. To address this, I created a [commit to address this in my personal fork](https://github.com/jack-stanek-hpe/sat/commit/a41b53f538db8bc7cd813763d67d8b576544a149). It prints a special error message when the `recipe` variable is undefined and `--recipe-version` isn't present, but returns the default error in all other cases.

Reply:
Neat to see an example of using the `undefined` kwarg 👍 

However, I should point out that `--recipe-version` is not actually what defines `recipe.version`. The `--recipe-version` flag will just tell `sat bootprep` which version of the recipe to go look up in VCS to find variables. Then there is a `recipe` key in the `product_vars.yaml` file. Maybe the recipe version should just come from the version of the recipe we specify or found in VCS based on existing branches. Also, I think all the other variables would give similarly confusing errors if you didn't have a `product_vars.yaml` file to look up.

I also originally wanted to make `--recipe-version` default to the latest version. I think that would resolve this issue in another way, would you agree?

Reply:
Yes, that would solve it in a different way, though the problem still exists in the case of the hpc-software-recipe repo not existing or if there's some error when fetching it -- in that case, we may not want to blanket fail in every case, since there's not theoretical reason that an input file which doesn't reference the recipe vars should fail if the recipe can't be fetched. In such a case, a "progressive enhancement" style paradigm could potentially suit both cases and cover the errors more gracefully, e.g. where the recipe vars are fetched on-demand when the recipe calls for them.

Reply:
Another thought I had w.r.t. variables being pulled in from the software recipe, which relates to [CRAYSAT-1530](CRAYSAT-1530), is that maybe it could be useful to dynamically document which variables are available given some context, in addition to what referential variables exist. For instance, given some recipe, the variables defined within that recipe could be listed out, perhaps using another subcommand.

Reply:
> Another thought I had w.r.t. variables being pulled in from the software recipe, which relates to [CRAYSAT-1530](CRAYSAT-1530), is that maybe it could be useful to dynamically document which variables are available given some context, in addition to what referential variables exist. For instance, given some recipe, the variables defined within that recipe could be listed out, perhaps using another subcommand.

Interesting idea. I opened [CRAYSAT-1532](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1532) for an option or action to list recipe variables.

Reply:
> Yes, that would solve it in a different way, though the problem still exists in the case of the hpc-software-recipe repo not existing or if there's some error when fetching it -- in that case, we may not want to blanket fail in every case, since there's not theoretical reason that an input file which doesn't reference the recipe vars should fail if the recipe can't be fetched. In such a case, a "progressive enhancement" style paradigm could potentially suit both cases and cover the errors more gracefully, e.g. where the recipe vars are fetched on-demand when the recipe calls for them.

I agree that it would be nice not to fail if the input file wasn't using the variables from the HPC software recipe. There's actually a comment in this method about changing an error to a warning when we default to the latest version of a recipe: https://github.com/Cray-HPE/sat/blob/9c509fa852e901efe60e2d11be4f5a5ec913130a/sat/cli/bootprep/vars.py#L115-L143

I think I had it implemented as a warning initially when I had it always defaulting to the latest software recipe version. Once I changed it so that the user had to specify a `--recipe-version`, I made it an error because in that case, the user would have had to explicitly ask for recipe variables.

Reply:
I'm going to change it to always attempt to load the latest HPC software recipe variables from VCS, and if it fails, I'm going to log a warning that rendering of templates which use those variables may fail.  That is, `--recipe-version` will not be required. I think that will improve usability significantly for the typical admin. Let me know if you disagree.

Being able to dynamically determine whether an input file is trying to use any recipe variables seems nice, but I'm not sure how we'd do that without knowing which variables we would expect a software recipe to define. I think it's probably good enough to at least default to the latest recipe version and only warn if there is a failure to load it.

Reply:
I made this change with 0b8fb2d6346d3dc1a0435a98636a7816116b4d3c.

I also filed [CRAYSAT-1533](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1533) to get the recipe version from the branch name rather than requiring it to be a variable in the file. This will require work here and in hpc-shasta-software-recipe.

Reply:
Oops, there's a bug in that commit which results in this:

```
ncn-m001:~/haasken/sprint-reviews/latest # sat bootprep run --dry-run --save-files --no-resolve-branches --recipe-version 22.07.2 bootprep-cos-config.yaml
INFO: Validating given input file bootprep-cos-config.yaml
INFO: Input file successfully validated against schema
WARNING: Failed to load variables defined by HPC software recipe version 22.07.2: No recipe with version 22.07.2 is installed on the system.
Traceback (most recent call last):
  File "/sat/venv/bin/sat", line 33, in <module>
    sys.exit(load_entry_point('sat==3.17.1', 'console_scripts', 'sat')())
  File "/sat/venv/lib/python3.9/site-packages/sat/main.py", line 120, in main
    subcommand(args)
  File "/sat/venv/lib/python3.9/site-packages/sat/cli/bootprep/main.py", line 266, in do_bootprep
    do_bootprep_run(schema_validator, args)
  File "/sat/venv/lib/python3.9/site-packages/sat/cli/bootprep/main.py", line 185, in do_bootprep_run
    var_context.load_vars()
  File "/sat/venv/lib/python3.9/site-packages/sat/cli/bootprep/vars.py", line 79, in load_vars
    deep_update_dict(self.vars, self.get_hpc_software_recipe_vars())
  File "/sat/venv/lib/python3.9/site-packages/sat/util.py", line 538, in deep_update_dict
    raise TypeError(f'Expected two dict instances. Got {type(original_dict)} and {type(new_dict)}')
TypeError: Expected two dict instances. Got <class 'dict'> and <class 'NoneType'>
```

I just need to return the empty dict in `get_hpc_software_recipe_vars` if there is an error loading the recipe vars.

Reply:
Pushed the fix b8ffde97553d3c3fdf086214dd3c1e83727669bc

New output for this test case:

```
(27dcf023d817) sat-container:/sat/share # sat bootprep run --dry-run --save-files --no-resolve-branches --recipe-version 22.07.2 bootprep-cos-config.yaml
INFO: Validating given input file bootprep-cos-config.yaml
INFO: Input file successfully validated against schema
WARNING: Failed to load variables defined by HPC software recipe version 22.07.2: No recipe with version 22.07.2 is installed on the system.
INFO: Would create 1 CFS configuration(s)
ERROR: Failed to render Jinja2 template haasken-cos-config-{{recipe.version}} for value name: 'recipe' is undefined
```